src/networkwatcher.cpp

```diff
@@ -0,0 +1,141 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "networkwatcher.h"
+#include "constants.h"
+#include "leakdetector.h"
+#include "logger.h"
+#include "mozillavpn.h"
+#include "networkwatcherimpl.h"
+#include "settingsholder.h"
+#include "systemtrayhandler.h"
+
+#ifdef MVPN_WINDOWS
+#  include "platforms/windows/windowsnetworkwatcher.h"
+#endif
+
+#ifdef MVPN_LINUX
+#  include "platforms/linux/linuxnetworkwatcher.h"
+#endif
+
+#ifdef MVPN_MACOS
+#  include "platforms/macos/macosnetworkwatcher.h"
+#endif
+
+#ifdef MVPN_WASM
+#  include "platforms/wasm/wasmnetworkwatcher.h"
+#endif
+
+// How often we notify the same unsecured network
+constexpr uint32_t NETWORK_WATCHER_TIMER_MSEC = 20000;
+
+namespace {
+Logger logger(LOG_NETWORKING, "NetworkWatcher");
+}
+
+NetworkWatcher::NetworkWatcher() { MVPN_COUNT_CTOR(NetworkWatcher); }
+
+NetworkWatcher::~NetworkWatcher() { MVPN_COUNT_DTOR(NetworkWatcher); }
+
+void NetworkWatcher::initialize() {
+  logger.log() << "Initialize";
+
+#if defined(MVPN_WINDOWS)
+  m_impl = new WindowsNetworkWatcher(this);
+#elif defined(MVPN_LINUX)
+  m_impl = new LinuxNetworkWatcher(this);
+#elif defined(MVPN_MACOS)
+  m_impl = new MacOSNetworkWatcher(this);
+#elif defined(MVPN_WASM)
+  m_impl = new WasmNetworkWatcher(this);
+#else
+  logger.log()
+      << "No NetworkWatcher implementation for the current platform (yet)";
+  return;
+#endif
+
+  connect(m_impl, &NetworkWatcherImpl::unsecuredNetwork, this,
+          &NetworkWatcher::unsecuredNetwork);
+
+  m_impl->initialize();
+
+  SettingsHolder* settingsHolder = SettingsHolder::instance();
+  Q_ASSERT(settingsHolder);
+
+  m_active = settingsHolder->unsecuredNetworkAlert();
+  if (m_active) {
+    m_impl->start();
+  }
+
+  connect(settingsHolder, &SettingsHolder::unsecuredNetworkAlertChanged, this,
+          &NetworkWatcher::settingsChanged);
+}
+
+void NetworkWatcher::settingsChanged(bool active) {
+  logger.log() << "Settings changed:" << active;
+  if (m_active == active) {
+    return;
+  }
+
+  m_active = active;
+
+  if (!m_impl) {
+    return;
+  }
+
+  if (m_active) {
+    m_impl->start();
+  } else {
+    m_impl->stop();
+  }
+}
+
+void NetworkWatcher::unsecuredNetwork(const QString& networkName,
+                                      const QString& networkId) {
+  logger.log() << "Unsecured network:" << networkName << "id:" << networkId;
+
+#ifndef UNIT_TEST
```

Comment:
Is this UNIT_TEST like unit testing?

Reply:
yes it is. For unit tests this part of the code is not needed.

Reply:
Fascinating. No rush, just more learning questions. (1) Is it normal to mark up main code with test notes like this? (2) Are there typical patterns where this happens?

Reply:
Our unit-tests are here: https://github.com/mozilla-mobile/mozilla-vpn-client/tree/main/tests/unit - We test only part of the codebase (yet) and, in order to do it, we compile part of the MozillaVPN code, together with the unit-test code.  When we want to disable/enable/change some detail, we use `#ifdef UNIT_TEST`.
We should reduce that at the minimum to have a better test coverage but there are a few exceptions:
1. MozillaVPN and ConnectionDataHolder - we use UNIT_TEST to have access to private class members:  exposing them would be a bad object-oriented logic encapsulation, but it's acceptable (and good for testing!) in unit-tests
2. Constants - we use UNIT_TEST to have faster operations such as: how often check new releases, how often check the IP addresses, etc etc. We have actually 3 values for most of the constants: production build, debug build, testing. See: contents.h
3. Logger - we use UNIT_TEST to compile and run android components even when not on android
4. VersionAPI, PingSender - we use UNIT_TEST to compile the Dummy ping sender or to check a generic update endpoint. Otherwise the unit-test would be platform specific.
5. Settings - we use UNIT_TEST because we don't want to write settings for real. The setting saving part is based on QSettings and I hope it's already tested by the QT framework

Here we do something similar to 4.